### PR TITLE
Add type matching benchmark

### PR DESCRIPTION
**Pull request by trask**
I added this to benchmark the change discussed in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/646#discussion_r451452916, and the benchmark seems useful to keep around (and probably improve).

### CODE REVIEW

File: benchmark/benchmark.gradle [hunk 1]
```diff
@@ -22,10 +25,16 @@ jmh {
   fork = 1 // How many times to forks a single benchmark. Use 0 to disable forking altogether
 //  jvmArgs += ["-XX:+UnlockDiagnosticVMOptions", "-XX:+DebugNonSafepoints", "-XX:StartFlightRecording=delay=5s,dumponexit=true,name=jmh-benchmark,filename=$rootDir/benchmark/build/reports/jmh/jmh-benchmark.jfr"]
 //  jvmArgs += ["-agentpath:$rootDir/benchmark/src/jmh/resources/libasyncProfiler.so=start,collapsed,file=$rootDir/benchmark/build/reports/jmh/profiler.txt"]
-  failOnError = true // Should JMH fail immediately if any benchmark had experienced the unrecoverable error?
+
+  // uncomment to use for TypeMatchingBenchmark
```
Comment (anuraaga) [inline benchmark/benchmark.gradle:29]: Is it worth having a separate Gradle project?

Comment (iNikem) [reply benchmark/benchmark.gradle:29]: Or separate Gradle task

File: benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+@BenchmarkMode(Mode.SingleShotTime)
+@Fork(5)
+@Warmup(iterations = 0)
+@Measurement(iterations = 1)
+@OutputTimeUnit(MILLISECONDS)
+@State(Scope.Thread)
+public class TypeMatchingBenchmark {
+
+  @Benchmark
+  public void loadLotsOfClasses() throws Exception {
```
Comment (iNikem) [inline benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java:44]: You would create a list of classes to load in `@Setup` method and benchmark itself will just load classes. And as a baseline, add a benchmark which just iterates over list of classes

**Comment by iNikem** [comment]
@trask do you want to address my comments and will merge?

**Comment by trask** [comment]
Yes, I do want to address your comments, they are good, just haven't got to it yet

File: benchmark/benchmark.gradle (hunk 1 above)
Comment (trask) [reply benchmark/benchmark.gradle]: I just tried this, but the jmh task seems different from normal gradle tasks(?): https://github.com/melix/jmh-gradle-plugin/blob/master/src/main/groovy/me/champeau/gradle/JMHTask.java#L32

and so I run into this problem:

https://github.com/melix/jmh-gradle-plugin/issues/161#issuecomment-662367084

Comment (trask) [reply benchmark/benchmark.gradle]: for now, I modeled it similar to existing benchmarks, and I added #782

File: benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    int errorCount = 0;
+    for (String className : classNames) {
+      try {
+        Class.forName(className, false, TypeMatchingBenchmark.class.getClassLoader());
+        successCount++;
+      } catch (Throwable t) {
+        errorCount++;
+      }
+    }
+    System.out.println("[loaded: " + successCount + ", failed to load: " + errorCount + "]");
```
Comment (anuraaga) [inline benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java:87]: Delete? Seems like it'd affect the benchmark result / be spammy

File: benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+@Warmup(iterations = 0)
+@Measurement(iterations = 1)
+@OutputTimeUnit(MILLISECONDS)
+@State(Scope.Thread)
+public class TypeMatchingBenchmark {
+
+  private Set<String> classNames;
+
+  @Setup
+  public void setup() throws IOException {
```
Comment (anuraaga) [inline benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java:50]: Ah just noticed maybe can stick with a static final. Find it makes benchmarks simpler not to worry about`State` unless really having to.

File: benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      try {
+        Class.forName(className, false, TypeMatchingBenchmark.class.getClassLoader());
+      } catch (NoClassDefFoundError e) {
+        // many classes in the jar files have optional dependencies which are not present
+      }
+    }
+  }
+
+  @Fork(jvmArgsAppend = "-javaagent:/path/to/opentelemetry-javaagent-master.jar")
+  public static class WithAgentMaster extends TypeMatchingBenchmark {}
```
Comment (iNikem) [inline benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java:81]: What is agent master?

File: benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+  @Fork(
+      jvmArgsAppend =
+          "-javaagent:/path/to/opentelemetry-java-instrumentation/java-agent/build/libs/opentelemetry-javaagent.jar")
```
Comment (iNikem) [inline benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java:85]: `java-agent`->`opentelemetry-agent`

File: benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java (hunk 5 above)
Comment (trask) [reply benchmark/src/jmh/java/io/opentelemetry/benchmark/TypeMatchingBenchmark.java]: i copied from another benchmark, i think the idea was to test against multiple versions of the agent, but i don't think needed so removed in both places

**Pull request merged**